client/client.go

```diff
@@ -137,47 +193,33 @@ func decodeHeader(header []byte) (code []byte, pt uint32, length int) {
 }
 
 func (client *Client) reconnect(err error) error {
-	if client.conn != nil {
-		return nil
-	}
 
-	// TODO I doubt this error handling is right because it looks
-	// really complicated.
+	// not actioning on error if it's deemed Temporary
+	// we might want to take note of timestamp and eventually recycle this connection
+	// if it persists too long (even though classified as Temporary here)
 	if opErr, ok := err.(*net.OpError); ok {
-		if opErr.Timeout() {
-			client.err(err)
-		}
 		if opErr.Temporary() {
 			return nil
 		}
-
-		return err
-	}
-
-	if err != nil {
-		client.err(err)
 	}
 
 	// If it is unexpected error and the connection wasn't
 	// closed by Gearmand, the client should close the conection
 	// and reconnect to job server.
-	client.Close()
+	log.Warningf("Closing connection to %v due to error %v, will reconnect...", client.addr, err)
 
-	conn, err := net.Dial(client.net, client.addr)
+	client.Lock()
+	if close_err := client.Close(); close_err != nil {
+		log.Warningf("Non-fatal error %v, while closing connection to %v", close_err, client.addr)
+	}
 
+	conn, err := client.handleConnOpen()
 	if err != nil {
-		client.err(err)
 		return err
 	}
 
-	swapped := atomic.CompareAndSwapPointer(
```

Comment:
Why would you replace CAS with a Mutex?

Reply:
CompareAndSwapPointer was removed only because I added use of lock(). Lock was added not because it's a substitute to CAS, instead the Lock() solution was chosen as a way to prevent the readLoop() and writeLoop() from using an old connection handle while the connection is being recovered in the reconnect() code initiated by either readLoop() or writeLoop(). As of the latest commit b19aec8, the other thread will wait for reconnection to complete and then proceed from there. This works in my testing so far, except for one kink with disconnect occurring right between sending command on a channel "outbound" and receiving a response on channel "expected"

Reply:
restored CAS with updated logic.